src/Workspaces/Core/Portable/Diagnostics/DiagnosticDataSerializer.cs

```diff
@@ -183,34 +185,42 @@ private static void WriteLocation(ObjectWriter writer, DiagnosticDataLocation? i
             writer.WriteInt32(item.MappedEndColumn);
         }
 
-        public ImmutableArray<DiagnosticData> ReadDiagnosticData(ObjectReader reader, Project project, TextDocument? document, CancellationToken cancellationToken)
+        public bool TryReadDiagnosticData(
+            ObjectReader reader,
+            Project project,
+            TextDocument? document,
+            CancellationToken cancellationToken,
+            out ImmutableArray<DiagnosticData> data)
         {
+            data = default;
+
             try
             {
                 var format = reader.ReadInt32();
                 if (format != FormatVersion)
                 {
-                    return default;
+                    return false;
                 }
 
                 // saved data is for same analyzer of different version of dll
                 var analyzerVersion = VersionStamp.ReadFrom(reader);
                 if (analyzerVersion != AnalyzerVersion)
                 {
-                    return default;
+                    return false;
                 }
 
                 var version = VersionStamp.ReadFrom(reader);
                 if (version != VersionStamp.Default && version != Version)
                 {
-                    return default;
+                    return false;
                 }
 
-                return ReadDiagnosticDataArray(reader, project, document, cancellationToken);
+                data = ReadDiagnosticDataArray(reader, project, document, cancellationToken);
+                return true;
             }
-            catch (Exception)
+            catch (Exception ex) when (FatalError.ReportWithoutCrashUnlessCanceled(ex))
             {
-                return default;
+                return false;
```

Comment:
this worries me in that it seems like we're testing a scenario where we write/read data, and are currently corrupting/crashing.  this feels like it papers over things.

Reply:
I agree, hence the reason to now log an NFW instead of silently swallowing the exception.

Reply:
I think we should rather let the exception propagate. I don't know why we even check versions here. Seems pointless.

---
In reply to: [466701364](https://github.com/dotnet/roslyn/pull/46597#discussion_r466701364) [](ancestors = 466701364)

Reply:
> I don't know why we even check versions here. Seems pointless. I'd change the above code to Contract.ThrowIfFalse and remove try-catch

@tmat `DiagnosticDataSerializer` is used in couple of code paths:
1. Serializing/deserializing diagnostics to/from devenv and OOP processes
2. Serializing/deserializing diagnostics within devenv to/from persistent storage. See the references in file `DiagnosticIncrementalAnalyzer.ProjectState.cs` here: http://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/Diagnostics/DiagnosticDataSerializer.cs,00debdf1025dbd8d,references.

For the second use case, it is very much possible/expected that the serialized versions do not match deserialized versions from persistent storage. So, the checks are valid. However, the current approach of continuing to attempt reading from the stream in presence of these version mismatches seem error-prone. Additionally, given there are multiple clients of DiagnosticDataSerializer, I don't feel comfortable letting the exceptions from core read part to propagate and have each client handle exceptions. My current approach stance is as follows:

1. Deserializing diagnostics from any serialized source is a safe operation with `TryDeserialize` semantics.
2. Each client that wants deserialized diagnostics is explicitly made aware about potential failures from the signature itself `bool TryXXX`, and we don't rely on each of them wrapping the deserialization attempts inside a try/catch

Reply:
For the scenario [2], where we need to check the version, shouldn't this check be only at the top-level though? That is have just one version for the whole diagnostic blob and then its parts are not versioned. Or is it valuable to be able to read partial data from the storage? 

I'm also wondering if versioning using a number is the best thing to do. If we make sure the next version is always additive and the extra data is ignored by previous reader then we would achieve more resiliency. Then we would only use the version number for truly breaking changes, which should be rare.

Reply:
Thanks @tmat - I filed https://github.com/dotnet/roslyn/issues/47801 for further refactoring work.